Bluetooth: Add a callback for notifying the previous bond state.

chejinxian opened the pull request:
bug: v/54141

HanJing755 reviewed (approved):
LGTM

JinxianChe commented on framework/include/bt_adapter.h:
```diff
@@ -363,6 +363,28 @@ void on_bond_state_changed(void* cookie, bt_address_t* addr, bt_transport_t tran
+ *     // Handle bond state change
+ * }
+ */
+typedef void (*on_bond_state_changed_callback_v2)(void* cookie, bt_address_t* addr, bt_transport_t transport, bond_state_t previous_state, bond_state_t current_state, bool is_ctkd);
```
parameter is enough？

JinxianChe commented on service/src/adapter_service.c:
```diff
@@ -2747,9 +2752,13 @@ bt_status_t adapter_pair_request_reply(bt_address_t* addr, bool accept)
         CALLBACK_FOREACH(CBLIST, adapter_callbacks_t, on_bond_state_changed,
             addr, BT_TRANSPORT_BREDR, BOND_STATE_BONDING, false);
+        CALLBACK_FOREACH(CBLIST, adapter_callbacks_t, on_bond_state_changed_v2,
+            addr, BT_TRANSPORT_BREDR, previous_state, BOND_STATE_BONDING, false);
```
1. initial value；
2. power on；
3. recommand v2 before v1.
4. notify bonding once or bonding to bonding.

JinxianChe replied:
LE remote cancel pair

JinxianChe commented on framework/include/bt_adapter.h:
```diff
@@ -512,6 +534,7 @@ typedef struct {
     on_connect_request_callback on_connect_request;
     on_connection_state_changed_callback on_connection_state_changed;
     on_bond_state_changed_callback on_bond_state_changed;
```
client notify twice in different callback

JinxianChe replied:
In order to be compatible with the way Local calls are made, the IPC message code cannot be reused.

JinxianChe replied:
No longer support this method in Loacl call.

gzh-terry commented on framework/include/bt_adapter.h:
```diff
@@ -363,6 +363,28 @@ void on_bond_state_changed(void* cookie, bt_address_t* addr, bt_transport_t tran
+ * @param transport - Transport type, see @ref bt_transport_t (0: LE, 1: BR/EDR).
+ * @param previous_state - Previous bond state, see @ref bond_state_t.
+ * @param current_state - Current bond state, see @ref bond_state_t.
+ * @param is_ctkd - Whether cross-transport key derivation is used.
```
`true` if cross-transport key derivation is used.

JinxianChe replied:
done

gzh-terry commented on framework/include/bt_adapter.h:
```diff
@@ -363,6 +363,28 @@ void on_bond_state_changed(void* cookie, bt_address_t* addr, bt_transport_t tran
+ * @param current_state - Current bond state, see @ref bond_state_t.
+ * @param is_ctkd - Whether cross-transport key derivation is used.
+ *
+ * **Example:**
```
An empty callback is not really informative. Can remove it.

JinxianChe commented on service/src/adapter_service.c:
```diff
@@ -2746,11 +2774,6 @@ bt_status_t adapter_pair_request_reply(bt_address_t* addr, bool accept)
     adapter_unlock();
     bt_status_t status;
     status = bt_sal_pair_reply(PRIMARY_ADAPTER, addr, accept ? 0 : HCI_ERR_PAIRING_NOT_ALLOWED);
-    if (status == BT_STATUS_SUCCESS && accept) {
```
keep

JinxianChe replied:
history

JinxianChe commented on service/src/adapter_service.c:
```diff
@@ -544,10 +544,16 @@ static void process_bond_state_change_evt(bt_address_t* addr, bond_state_t state
 {
     remote_device_properties_t remote;
     bt_device_t* device;
+    bond_state_t old_state;
```
definition

JinxianChe replied:
Temporarily abandon the idea of synchronizing with the protocol stack. Choose to determine whether to report the status at the location where the binding state is set, in order to avoid introducing too many modifications.

chengkai15 commented on service/src/adapter_service.c:
```diff
@@ -2715,6 +2745,8 @@ bt_status_t adapter_remove_bond(bt_address_t* addr, uint8_t transport)
 
     adapter_unlock();
+    // Notify bonding state change
+    CALLBACK_FOREACH(CBLIST, adapter_callbacks_t, on_bond_state_changed_v2, addr, transport, state, BOND_STATE_NONE, false);
```
1. when it was  no bond,  on_bond_state_changed_v2 would callback none state.

chengkai15 commented on service/src/adapter_service.c:
```diff
@@ -1598,6 +1620,13 @@ void adapter_init(void)
 
 void* adapter_register_callback(void* remote, const adapter_callbacks_t* adapter_cbs)
 {
+#ifdef CONFIG_BLUETOOTH_FRAMEWORK_LOCAL
```
remove? BT_CALLBACK_FOREACH would check on_bond_state_changed cb

chengkai15 commented on service/src/adapter_service.c:
```diff
@@ -569,15 +580,19 @@ static void process_bond_state_change_evt(bt_address_t* addr, bond_state_t state
 #endif
     }
 
     device_set_bond_state(device, state);
```
if we callback on_bond_state_changed_v2 within device_set_bond_state, which maybe easy?

let's  follow BondStateMachine.java in android

JinxianChe replied:
do in service loop?
ctkd field

JinxianChe commented on service/ipc/socket/src/bt_socket_adapter.c:
```diff
@@ -587,11 +588,20 @@ int bt_socket_client_adapter_callback(service_poll_t* poll,
     }
     case BT_ADAPTER_ON_BOND_STATE_CHANGED: {
+        CALLBACK_FOREACH(CBLIST, adapter_callbacks_t,
+            on_bond_state_changed_v2,
```
rename.

JinxianChe replied:
on_bond_state_changed_extra

gzh-terry commented on service/src/adapter_service.c:
```diff
@@ -544,10 +544,16 @@ static void process_bond_state_change_evt(bt_address_t* addr, bond_state_t state
 {
     remote_device_properties_t remote;
     bt_device_t* device;
+    bond_state_t old_state;
```
Can name this `prev_state`

gzh-terry commented on service/src/adapter_service.c:
```diff
@@ -638,15 +653,22 @@ static void process_link_key_update_evt(bt_address_t* addr, bt_128key_t link_key
+
+    /* notify bond state change */
+    if (state == BOND_STATE_BONDED) {
+        CALLBACK_FOREACH(CBLIST, adapter_callbacks_t, on_bond_state_changed_v2, addr, BT_TRANSPORT_BREDR, state, BOND_STATE_NONE, false);
```
This is a compensation of `device_set_bond_state(device, BOND_STATE_NONE);` in line 663.

chejinxian committed:
API: Add a new callback interface to obtain bond states.

bug: v/54141

Rootcause: The new interface will inform the upper-layer application of the previous bond state, thus distinguishing the different scenarios that may occur during the bonding process.

Signed-off-by: chejinxian1 <chejinxian1@xiaomi.com>

chejinxian committed:
Adapter: Modify bond state notification method.

bug: v/54141

Rootcause: Modify the implementation of the bond state notification to support reporting the previous state. To more accurately report the bond state at the Framework layer, this modification will trigger a notification event when changing the bond state. To avoid long-term occupation of the adapter_lock, the method of do_in_service_loop is used to execute the actual notification.

Signed-off-by: chejinxian1 <chejinxian1@xiaomi.com>

chejinxian committed:
Adapter: Modify the timing of entering bonding state.

bug: v/54141

Rootcause: After agreeing to exchange IO capabilities, the upper layer was notified to enter the bonding state, but the Framework did not actually enter bonding. The purpose of entering bonding here is to avoid the situation where insufficient IO capabilities prevent the corresponding information from being obtained from the protocol stack to enter bonding.

Signed-off-by: chejinxian1 <chejinxian1@xiaomi.com>

chejinxian committed:
bttool: Adapt to the latest bond status callback interface.

bug: v/54141

Rootcause: Modify the callback function for obtaining the bond state in bttool.

Signed-off-by: chejinxian1 <chejinxian1@xiaomi.com>

chejinxian committed:
Adapter: Cancel reporting BOND_STATE_CANCELING.

bug: v/54141

Rootcause: BOND_STATE_CANCELING is the transitional state from BOND_STATE_BONDING to BOND_STATE_BONDED in the case of actively canceling the pairing. Since only the upper layer initiates the pairing cancellation will enter this state, the upper layer application does not care about entering BOND_STATE_CANCELING, so filter out this state report. This is more in line with the habits of Bluetooth software development.

Signed-off-by: chejinxian1 <chejinxian1@xiaomi.com>

chengkai15 commented on service/src/adapter_service.c:
```diff
@@ -118,6 +118,41 @@ static void adapter_unlock(void)
+
+    CALLBACK_FOREACH(CBLIST, adapter_callbacks_t, on_bond_state_changed_extra, addr, transport,
+        previous_state, current_state, msg->is_ctkd);
+    free(msg);
```
msg maybe memory Leaks,  handle it within LINE:145

JinxianChe replied:
done

gzh-terry commented on service/src/adapter_service.c:
```diff
@@ -118,6 +118,42 @@ static void adapter_unlock(void)
+        }
+    }
+
+    CALLBACK_FOREACH(CBLIST, adapter_callbacks_t, on_bond_state_changed_extra, addr, transport,
```
Will this be triggered if the app only register the legacy `on_bond_state_changed` callback?
